### PR TITLE
Language model

### PR DESCRIPTION
**Pull request by chengxiaohua1105**
train a language model by ngrm

### CODE REVIEW

File: language_model/ngram_language_model.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+########################################################################
+# 
+# Copyright (c) 2017 Baidu.com, Inc. All Rights Reserved
+# 
+########################################################################
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: Please remove the license.

File: language_model/ngram_language_model.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+"""
+File: ngram_language_model.py
+Author: chengxiaohua(chengxiaohua@baidu.com)
+Date: 2017/05/02 15:13:21
+"""
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: remove line 8 ~ 12.

File: language_model/ngram_language_model.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import sqlite3
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: remove line 13, because it is not implicitly used.

File: language_model/ngram_language_model.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+batch_size = 1000
+word_dim = 256
+hidden_dim = 512
+lr = 0.1
+max_word_num = 40
+
+
+def gen_embedding_layer(layer_name, voc_dim):
+    return paddle.layer.data(
+        name=layer_name, type=paddle.data_type.integer_value(voc_dim))
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: `gen_embedding_layer` is not necessary and the name is confusing. I suggest to directly call the data layer.

File: language_model/ngram_language_model.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def word_embed(in_layer):
+    word_embed = paddle.layer.table_projection(
+        input=in_layer,
+        size=word_dim,
+        param_attr=paddle.attr.Param(
+            name="emb", initial_std=0.001, learning_rate=lr, l2_rate=0))
+    return word_embed
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: line 34 ~ 40, please directly call `paddle.layer.embedding`.

File: language_model/ngram_language_model.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+
+def ngrm_network(voc_dim):
+    firWord = gen_embedding_layer("firWord", voc_dim)
+    secWord = gen_embedding_layer("secWord", voc_dim)
+    thirdWord = gen_embedding_layer("thirdWord", voc_dim)
+    forthWord = gen_embedding_layer("forthWord", voc_dim)
+    fifthWord = gen_embedding_layer("fifthWord", voc_dim)
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: line 44 ~ 48, first_word, second_word, third_word, fourth_word, fifth_word。please use all lower cases.

File: language_model/ngram_language_model.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+    fir = word_embed(firWord)
+    sec = word_embed(secWord)
+    third = word_embed(thirdWord)
+    forth = word_embed(forthWord)
+    fifth = word_embed(firWord)
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: line 50 ~ 54, word_embed --> word_embedding

File: language_model/ngram_language_model.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+    embedding = paddle.layer.concat(input=[fir, sec, third, forth, fifth])
+
+    hidden1 = paddle.layer.fc(
+        input=embedding,
+        act=paddle.activation.Sigmoid(),
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: change the activation from sigmoid to tanh.

File: language_model/ngram_language_model.py [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+        gradient_clipping_threshold=25)
+
+    trainer = paddle.trainer.SGD(cost, parameters, adadelta_optimizer)
+    trainer.train(
+        paddle.batch(ptb.ngram_train(word_dict, 6), batch_size),
+        num_passes=30,
+        event_handler=event_handler)
+
+
+def predict(word_dict, voc_dim):
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: predict --> generate_sequence.

By using a language model to generate the sequence, it should be provided the first sentence. Here, how to get the input sentence?

File: language_model/ngram_language_model.py [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+    prediction_layer = ngrm_network(voc_dim)
+
+    with gzip.open('./models/model_pass_00000.tar.gz') as f:
+        parameters = paddle.parameters.Parameters.from_tar(f)
+
+    idx_word_dict = dict((v, k) for k, v in word_dict.items())
+    ins_iter = paddle.dataset.imikolov.test(word_dict, 5)
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py:128]: Please give some comments to explain how to generate sequences.Tell the users this example only implements one-way search sample-by-sample and suggest them implementing beam search.

File: language_model/ngram_language_model.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+    for ins in ins_iter():
+        gen_seq = []
+        cur_predict_word = None
+        pre_n_words = [ins[-i] for i in xrange(4, 0, -1)]
+        while len(gen_seq) < max_word_num:
+            infer_res = paddle.infer(
+                output_layer=prediction_layer,
+                parameters=parameters,
+                input=[pre_n_words])
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: Please use a maxid layer in the generation.

File: language_model/ngram_language_model.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+def process(is_generating):
+    paddle.init(use_gpu=False, trainer_count=4)
+    word_dict = paddle.dataset.imikolov.build_dict()
+    voc_dim = len(word_dict)
+
+    if not is_generating:
+        train_model(word_dict, voc_dim)
+    elif is_generating:
+        predict(word_dict, voc_dim)
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: The name `predict` is in-appropriate. This is not predicting but generating.

**Review by lcy-seso — COMMENTED** [review]
The codes need to be modified.

**Review by lcy-seso — CHANGES_REQUESTED** [review]
Please refine codes of generating the sequence.

File: language_model/ngram_language_model.py (hunk 1 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 2 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 3 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 4 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 5 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 6 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 7 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 8 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 12 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py (hunk 9 above)
Comment (chengxiaohua1105) [reply language_model/ngram_language_model.py]: done

File: language_model/ngram_language_model.py [hunk 13]
```diff
@@ -107,13 +96,24 @@ def event_handler(event):
...
+        paddle.batch(paddle.dataset.imikolov.train(word_dict, 6), batch_size),
         num_passes=30,
         event_handler=event_handler)
 
 
-def predict(word_dict, voc_dim):
-    prediction_layer = ngrm_network(voc_dim)
+def generate_sequence(word_dict, voc_dim):
+    """
+    the generating process as follows:
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: the generating process as follows --> the generating process goes as follows

File: language_model/ngram_language_model.py [hunk 14]
```diff
@@ -107,13 +96,24 @@ def event_handler(event):
...
+      1 construct the ngram netword
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: What does netword mean？

File: language_model/ngram_language_model.py [hunk 15]
```diff
@@ -107,13 +96,24 @@ def event_handler(event):
...
+      2 read the model which is trained before
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: read the model which is trained before --> load the trained model.

File: language_model/ngram_language_model.py [hunk 16]
```diff
@@ -107,13 +96,24 @@ def event_handler(event):
...
+      3 read the test data by ins_iter which is a five-element couple
+      4 generate the next word according to the last four words
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: How to generate the next word? please explain the idea.

File: language_model/ngram_language_model.py [hunk 17]
```diff
@@ -107,13 +96,24 @@ def event_handler(event):
...
+    the generating process as follows:
+      1 construct the ngram netword
+      2 read the model which is trained before
+      3 read the test data by ins_iter which is a five-element couple
+      4 generate the next word according to the last four words 
+          until the word count is equal to max_word_num or 
+          the next word is <e>
+      5 so far, we have completed one five-element couple predicting,
+        and then continue to iterate the next couple
+    """
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: The explanation is expected to explain the idea of the one-way search algorithm.

File: language_model/ngram_language_model.py [hunk 18]
```diff
@@ -75,14 +58,19 @@ def ngrm_network(voc_dim):
 
 
 def train_model(word_dict, voc_dim):
-    nxtWord = gen_embedding_layer("nxtWord", voc_dim)
-    output = ngrm_network(voc_dim)
+    #treate the next_word as label
+    nxtWord = paddle.layer.data(
```
Comment (lcy-seso) [inline language_model/ngram_language_model.py]: nxtWord --> next_word. please make sure the naming style consistent.

File: language_model/ngram_language_model.py (hunk 17 above)
Comment (lcy-seso) [reply language_model/ngram_language_model.py]: How can generating sequence without input?

**Review by lcy-seso — CHANGES_REQUESTED** [review]
Please modify all the comments before a new code review.
